types/qs/index.d.ts

```diff
@@ -55,6 +55,16 @@ declare namespace QueryString {
         interpretNumericEntities?: boolean;
     }
 
+    interface ParsedQs { [key: string]: string | string[] | ParsedQs | ParsedQs[] }
+
+    // The value type here is a "poor man's `unknown`". When these types support TypeScript
+    // 3.0+, we can replace this with `unknown`.
+    type PoorMansUnknown = {} | null | undefined;
+
+    // Not supported natively until TS 3.5
+    type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
+
     function stringify(obj: any, options?: IStringifyOptions): string;
-    function parse(str: string, options?: IParseOptions): any;
+    function parse(str: string, options?: Omit<IParseOptions, 'decoder'>): ParsedQs;
+    function parse(str: string, options?: IParseOptions): { [key: string]: PoorMansUnknown };
```

Comment:
i wonder if we could infer, from the return type of the decoder option, what the value of this object might be?

Reply:
We sort of discussed this earlier, here: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/43440#discussion_r399768952.

I can look into it again though. Is it possible to write a decoder that creates a key with a type other than string, or is the key always a string (at runtime)?

Reply:
yes, `parse`'s decoder can return any value. however, if it's for a key, it'll be whatever the engine allows as an object key - which at the moment, in TS, would be `string || symbol || number || bigint` (which should be `PropertyKey` but i don't think that includes all the proper primitive types)

Reply:
Unless I’m missing something, I don’t think these overloads make sense. It seems like you’re trying to return a more specific type (`ParsedQs`) if you know that you _weren’t_ passed a `decoder`. But, since `IParseOptions` is assingable to `Omit<IParseOptions, 'decoder'>`, you can’t really know for sure that you weren’t passed a decoder.

[This](https://www.typescriptlang.org/play/#code/C4TwDgpgBAJhBmBDArgG2AUQHYGMD2cATlALxQAUAzsIQFxSJYgA0sCK62+RA-PYyyg4AFokKUIwPlGqEAllgDmASlIA+GTQWKA3AChQkNkjTAAIhG4RiZKjXqztrOFcLSBrEWIlSHWpaokGo5K+noKwNZIONAAkgAK3hAA8mDAcnhYlFAA3npQBUJ4ALbFiNIARnh4qBCM+oVsqHLFcpFufvJKUAA+UABKEIoYAB5gDYVwacLSWMjFFda9UEioEhMFLgTW0nZ0ml2KzuymFq70cCboZ9uEnqLikp1OUIYQ9ADkANYQIB-LHwAbohUMgIB9AhoBBsGIRCIgQAAZFptWbzRaEGFgJIAQThCMolWqtXq+UKINQeAA7mY8MBCfQqjU6lgsahEApkhUAFaWelE5mkxoU6nxQh0ungCAMqBMkmsskFbHw4qSazI1q+KBzBbWGGOHDAAByaFQAAlGDBmkoBfKYXJFFg8IQIABFMGEEBihByEa2lkwryPLUfZDAeAAWgAHP8+h85JQ8NGowBWACcEYAjB9Aw8fABlCBYdJYCCof1CwoRaxgF3G+bWOQ4bDpdLSisKgC+enCxaiiBiUESjxgruyOSgAG0fiBnkoALpzxTLEKKSfz5bDiSj7J9LcQHfrqDdgxSofVQgAWUYlAAqlgvk6qVhSLlO8s5qhUMtkFhLgoDzCFx2RdFZf0NDIX2VCQ9iXVg8DSSCZWSTUAB4EiSVJ0kyShWA+LYiA+NRlHofcd30YCxGgeBwOwqCklgg4XgQuiZQwx4sKQkjcinGcl0Xc9nWvLJ70fakX07MJ8CyYAoBYpD6HYiROJw188kaAjrHochISgAAGPRJJ7aTqCgABHbIyGgiByA+D54MQnDlH0IA) demonstrates, I think, a case where my decoder unconditionally returns a number, but these typings think it’s safe to assume that all properties are `string | string[] | ParsedQs | ParsedQs[]`.

Reply:
Good catch. How about this?

```diff
-declare function parse(str: string, options?: Omit<IParseOptions, 'decoder'>): ParsedQs;
+declare function parse(str: string, options?: Omit<IParseOptions, 'decoder'> & Required<Pick<IParseOptions, 'decoder'>>): ParsedQs;
 declare function parse(str: string, options?: IParseOptions): { [key: string]: PoorMansUnknown };
```

https://www.typescriptlang.org/play/?ssl=26&ssc=1&pln=27&pc=1#code/C4TwDgpgBAJhBmBDArgG2AUQHYGMD2cATlALxQAUAzsIQFxSJYgA0sCK62+RA-PYyyg4AFokKUIwPlGqEAllgDmASlIA+GTQWKA3AChQkNkjTAAIhG4RiZKjXqztrOFcLSBrEWIlSHWpaokGo5K+noKwNZIONAAkgAK3hAA8mDAcnhYlFAA3npQBUJ4ALbFiNIARnh4qBCM+oVsqHLFcpFufvJKUAA+UABKEIoYAB5gDYVwacLSWMjFFda9UEioEhMFLgTW0nZ0ml2KzuymFq70cCboZ9uEnqLikp1OUIYQ9ADkANYQIB-LHwAbohUMgIB9AhoBBsGIRCIgQAAZFptWbzRaEGFgJIAQThCMolWqtXq+UKINQeAA7mY8MBCfQqjU6lgsahEApkhUAFaWelE5mkxoU6nxQh0ungCAMqBMkmsskFbHw4qSazI1q+KBzBbWGGOHDAAByaFQAAlGDBmkoBfKYXJFFg8IQIABFMGEEBihByEa2lkwryPLUfZDAeAAWgAHP8+h85JQ8NGowBWACcEYAjB9Aw8fABlCBYdJYCCof1CwoRaxgF3G+bWOQ4bDpdLSisKgC+enCxaiiBiUESjxgruyOSgAG0fiBnkoALpzxTLEKKSfz5bDiSj7J9LcQHfrqDdgxSofVQgAWUYlAAqlgvk6qVhSLlO8s5qhUMtkFhLgoDzCFx2RdFZf0NDIX2VCQ9iXVg8DSSCZWSTUAB4EiSVJ0kyShWA+LYiA+DQADIBggABHZA5BdGBUPiJsvnQ-csKQvCCOsIi1GUeh9x3fRgLEaB4HA7CoKSWCDheBDRJlDDHhYnDuNyKcZyXRdz2da8snvR9qRfTswnwLJgCgaSkPoOSJAUrJXzyRp2P2chISgAAGPQDJ7IzqCgcjsjIaCIHID4PngxDFP0LyTN8gAmV8AqCkLckVNhzgoZy3M7ZR9CAA

Reply:
Yeah, I think that works. I think with that you don’t need the `Omit` anymore.

Reply:
Updated the PR. Would you mind giving it another review?